python/ray/external_storage.py

```diff
@@ -90,6 +129,223 @@ def restore_spilled_objects(self, keys):
+            key = self.prefix + ref.hex()
+            self._spill_object(key, ref, buf, metadata)
+            keys.append(key.encode())
+        return keys
+
+    def restore_spilled_objects(self, keys):
+        for k in keys:
+            key = k.decode()
+            ref = ray.ObjectRef(bytes.fromhex(key[len(self.prefix):]))
+            self._restore_spilled_object(key, ref)
+
+    def get_stats(self):
+        # TODO(sang): Implement it for benchmark.
+        return {}
+
+    def _spill_object(self, key, ref, buf, metadata):
+        s3_object = self.s3.Object(self.bucket_name, key)
+
+        with BytesIO() as file_like:
+            metadata_len = len(metadata)
+            buf_len = len(buf)
+            file_like.write(metadata_len.to_bytes(8, byteorder="little"))
+            file_like.write(buf_len.to_bytes(8, byteorder="little"))
+            file_like.write(metadata)
+            file_like.write(memoryview(buf))
+            file_like.seek(0)
+            # TODO(sang): Should we check integrity here?
+            # TODO(sang): Add a job id to a tagging field for deletion.
+            s3_object.put(Body=file_like)
+
+    def _restore_spilled_object(self, key, ref):
+        resp = self.s3.Object(self.bucket_name, key).get()
+        if "Body" not in resp:
+            raise IOError("Failed to get a proper response from S3. "
+                          "The response should contain a Body field, "
+                          f"but it only got {resp}")
+        with closing(resp["Body"]) as streaming_body:
+            # botocore.response.StreamingBody doesn't provide readinto,
+            # which we'd need to put objects to avoid unnecessary copy.
+            # Solution 1: Use BytesIO and pass bytes using
+            #   streaming_body.read(). This solution requires a byte copy
+            #   returned by the read API.
+            # Solution 2: Use underlying _raw_stream attribute that is
+            #   a true file-like object. This seems to be fine because
+            #   StreamingBody class seems to be a simple wrapper around
+            #   this raw stream. Look at source code;
+            #   https://gemfury.com/squarecapadmin/python:botocore/-/content/botocore/response.py # noqa
+            #   BufferedReader allows us to copy underlying bytes
+            #   directly to Ray object buffer.
+            # We are going with solution 2 to reduce memory footprint.
+            with BufferedReader(streaming_body._raw_stream) as file_like:
+                metadata_len = int.from_bytes(
+                    file_like.read(8), byteorder="little")
+                buf_len = int.from_bytes(file_like.read(8), byteorder="little")
+                metadata = file_like.read(metadata_len)
+                # read remaining data to our buffer
+                self._put_object_to_store(metadata, buf_len, file_like, ref)
+
+
+class S3StorageSmartOpenImpl(ExternalStorage):
```

Comment:
```suggestion
class ExternalStorageSmartOpenImpl(ExternalStorage):
```

smart_open works with many storage systems; so we should not include "S3" in the name here.

Reply:
Ah, I see. So you suggest making this implementation completely independent from storages. That sounds like a good idea. I will address all the comments below.